include/deal.II/lac/tensor_product_matrix.h

```diff
@@ -838,6 +1037,153 @@ TensorProductMatrixSymmetricSum<dim, Number, n_rows_1d>::reinit(
 
 
 
+template <int dim, typename Number, int n_rows_1d>
+void
+TensorProductMatrixSymmetricSumCache<dim, Number, n_rows_1d>::reserve(
+  const unsigned int size)
+{
+  indices.assign(size * dim, numbers::invalid_unsigned_int);
+}
+
+
+
+template <int dim, typename Number, int n_rows_1d>
+void
+TensorProductMatrixSymmetricSumCache<dim, Number, n_rows_1d>::insert(
+  const unsigned int                       index,
+  const std::array<Table<2, Number>, dim> &Ms,
+  const std::array<Table<2, Number>, dim> &Ks)
+{
+  for (unsigned int d = 0; d < dim; ++d)
+    {
+      const MatrixPairType matrix(Ms[d], Ks[d]);
+
+      const auto ptr = cache.find(matrix);
+
+      if (ptr != cache.end())
+        indices[index * dim + d] = ptr->second;
+      else
+        {
+          indices[index * dim + d] = cache.size();
+          cache[matrix]            = cache.size();
+        }
+    }
+}
+
+
+
+template <int dim, typename Number, int n_rows_1d>
+void
+TensorProductMatrixSymmetricSumCache<dim, Number, n_rows_1d>::finalize()
+{
+  vector_mass_matrix.resize(cache.size());
+  vector_derivative_matrix.resize(cache.size());
+  vector_eigenvectors.resize(cache.size());
+  vector_eigenvalues.resize(cache.size());
+
+  const auto store = [&](const unsigned int    index,
+                         const MatrixPairType &M_and_K) {
+    std::array<Table<2, Number>, 1> mass_matrices;
+    mass_matrices[0] = M_and_K.first;
+
+    std::array<Table<2, Number>, 1> derivative_matrices;
+    derivative_matrices[0] = M_and_K.second;
+
+    std::array<Table<2, Number>, 1>      eigenvectors;
+    std::array<AlignedVector<Number>, 1> eigenvalues;
+
+    internal::TensorProductMatrixSymmetricSum::setup(mass_matrices,
+                                                     derivative_matrices,
+                                                     eigenvectors,
+                                                     eigenvalues);
+
+    vector_mass_matrix[index]       = M_and_K.first;
+    vector_derivative_matrix[index] = M_and_K.second;
+    vector_eigenvectors[index]      = eigenvectors[0];
+    vector_eigenvalues[index]       = eigenvalues[0];
+  };
+
+  if (cache.size() == indices.size())
+    {
+      std::map<unsigned int, MatrixPairType> inverted_cache;
+
+      for (const auto &i : cache)
+        inverted_cache[i.second] = i.first;
+
+      for (unsigned int i = 0; i < indices.size(); ++i)
+        store(i, inverted_cache[indices[i]]);
+
+      indices.clear();
+    }
+  else
+    {
+      for (const auto &i : cache)
+        store(i.second, i.first);
+    }
```

Comment:
Do you really need two different code paths (and the associated lambda function)? It is only a matter of the index you are using, and I would intuitively have said that we need to store at `i.second` also in that case, because we should have `ptr->second == indices[index * dim + d]` according to the assignment in `insert`.